src/aiida_wannier90_workflows/utils/pseudo/__init__.py

```diff
@@ -259,20 +405,68 @@ def get_number_of_projections(
     for site in structure.sites:
         upf = pseudos[site.kind_name]
         nprojs = get_number_of_projections_from_upf(upf)
-        soc = is_soc_pseudo(get_upf_content(pseudos[site.kind_name]))
-        if spin_orbit_coupling and not soc:
-            # For SOC calculation with non-SOC pseudo, QE will generate
+        if spin_non_collinear and not spin_orbit_coupling:
+            # For magnetic calculation with non-SOC pseudo, QE will generate
             # 2 PSWFCs from each one PSWFC in the pseudo
+            # For collinear-magnetic calculation, spin up and down will calc
+            # seperately, so nprojs do not times 2
             nprojs *= 2
-        elif not spin_orbit_coupling and soc:
-            # For non-SOC calculation with SOC pseudo, QE will average
+        elif not spin_non_collinear and spin_orbit_coupling:
+            # For non-magnetic calculation with SOC pseudo, QE will average
             # the 2 PSWFCs into one
             nprojs //= 2
         tot_nprojs += nprojs
 
     return tot_nprojs
 
 
+def get_number_of_projections_ext(
+    structure: orm.StructureData,
+    external_projectors: dict,
+    spin_non_collinear: bool,
+    spin_orbit_coupling: bool = False,
+) -> int:
+    """Get number of projections for the structure with the given projector dict.
+
+    :param structure: crystal structure
+    :type structure: aiida.orm.StructureData
+    :param projectors: a dictionary contains projector list of the structure
+    :type pseudos: dict
+    :param spin_non_collinear: non-collinear or spin-orbit-coupling
+    :type spin_non_collinear: bool
+    :return: number of projections
+    :rtype: int
+    """
+    if not isinstance(structure, orm.StructureData):
+        raise ValueError(
+            f"The type of structure is {type(structure)}, only aiida.orm.StructureData is accepted"
+        )
+    # e.g. composition = {'Ga': 1, 'As': 1}
+    composition = structure.get_composition()
+
+    # I use the first projector to detect SOCs
+    kind = list(composition.keys())[0]
+    spin_orbit_coupling_proj = "j" in external_projectors[kind][0]
+    if spin_orbit_coupling and not spin_orbit_coupling_proj:
+        raise ValueError("Use SOC spin type but non-SOC projectors.")
+    if spin_orbit_coupling_proj and not spin_orbit_coupling:
+        raise ValueError("Use SOC projectors but non-SOC spin type.")
+
+    tot_nprojs = 0
+    for kind in composition:
+        nprojs = 0
+        for orb in external_projectors[kind]:
+            if spin_orbit_coupling:
+                nprojs += round(2 * orb["j"]) + 1
+            else:
+                nprojs += 2 * orb["l"] + 1
+        if spin_non_collinear and not spin_orbit_coupling:
+            nprojs *= 2
+        tot_nprojs += nprojs * composition[kind]
```

Comment:
This loop iterates through atom symbols that `structure.get_composition()` returns and does not account for changes in #60. However, I'm not 100% sure if it's required in the case of external projectors.
If accounting for different kind names, the loop should be over `structure.sites`:
```suggestion
    for site in structure.sites:
        nprojs = 0
        for orb in external_projectors[site.kind_name]:
            if spin_orbit_coupling:
                nprojs += round(2 * orb["j"]) + 1
            else:
                nprojs += 2 * orb["l"] + 1
        if spin_non_collinear and not spin_orbit_coupling:
            nprojs *= 2
        tot_nprojs += nprojs * composition[kind]
```